File: 백준/Gold/2493. 탑/탑.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const N = Number(input[0])
const heights = input[1].split(" ").map(Number)
const answer = new Array(N).fill(0)
const stack = []


for(let i = N-1;i>=0;i--){
  // stack과 heights 비교
  while(stack.length>0 && heights[stack[stack.length-1]] <= heights[i]){
    const targetIndex = stack.pop()
    answer[targetIndex] = i+1
  }

  stack.push(i) // index 추가
}

console.log(answer.join(' '))

File: 백준/Gold/2493. 탑/탑.py
import sys
input = sys.stdin.readline

N = int(input())
heights = list(map(int,input().split()))
stack = []
result = [str(0) for _ in range(N)]

for i in range(N-1,-1,-1):
  while len(stack) >0 and heights[stack[-1]] <= heights[i]:
    cur = stack.pop()
    result[cur] = str(i+1)
  stack.append(i)

print(' '.join(result))